test/powershell/Modules/Microsoft.PowerShell.Utility/Foreach-Object-Parallel.Tests.ps1

```diff
@@ -0,0 +1,330 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+Describe 'ForEach-Object -Parallel Basic Tests' -Tags 'CI' {
+
+    BeforeAll {
+
+        $skipTest = -not $EnabledExperimentalFeatures.Contains('PSForEachObjectParallel')
+        if ($skipTest) {
+            Write-Verbose "Test Suite Skipped. The test suite requires the experimental feature 'PSForEachObjectParallel' to be enabled." -Verbose
+            $originalDefaultParameterValues = $PSDefaultParameterValues.Clone()
+            $PSDefaultParameterValues["it:skip"] = $true
+        }
+        else {
+            $sb = { "Hello!" }
+        }
+    }
+
+    AfterAll {
+
+        if ($skipTest) {
+            $global:PSDefaultParameterValues = $originalDefaultParameterValues
+        }
+    }
+
+    It "Verifies dollar underbar variable" {
+
+        $expected = 1..10
+        $result = $expected | ForEach-Object -Parallel { $_ }
+        $result.Count | Should -BeExactly $expected.Count
+        $result | Should -Contain 1
+        $result | Should -Contain 10
+    }
+
+    It 'Verifies using variables' {
+
+        $var = "Hello"
+        $varArray = "Hello","There"
+        $result = 1..1 | ForEach-Object -Parallel { $using:var; $using:varArray[1] }
+        $result.Count | Should -BeExactly 2
+        $result[0] | Should -BeExactly $var
+        $result[1] | Should -BeExactly $varArray[1]
+    }
+
+    It 'Verifies terminating error streaming' {
+
+        $result = 1..1 | ForEach-Object -Parallel { throw 'Terminating Error!'; "Hello" } 2>&1
+        $result.Count | Should -BeExactly 1
+        $result.ToString() | Should -BeExactly 'Terminating Error!'
+        $result.FullyQualifiedErrorId | Should -BeExactly 'PSTaskException'
+    }
+
+    It 'Verifies terminating error in multiple iterations' {
+
+        $results = 1..2 | ForEach-Object -Parallel {
+            if ($_ -eq 1) {
+                throw 'Terminating Error!'
+                "Hello!"
+            }
+            else {
+                "Goodbye!"
+            }
+        } 2>&1
+
+        $results | Should -Not -Contain "Hello!"
+        $results | Should -Contain "Goodbye!"
+    }
+
```

Comment:
can we add tests for `ErrorAction` and `ErrorVariable`

Reply:
ErrorAction and ErrorVariable is not supported for this parameter set, in that they will not affect what each parallel execution does, e.g., ErrorAction will not terminate a running script block as all terminating errors are caught and written to the pipeline.  Similarly ErrorVariable will not collect errors from each running script block.

However, it is something we can look in the future.

Reply:
Since ErrorAction, ErrorVariable, WarningAction, etc. parameters are not currently supported I'll add code to throw when used for now.  In the future we should look at adding support for ErrorVariable, WarningVariable, etc.

Thanks!